router/batchrouter/partition_worker.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/rudderlabs/rudder-go-kit/stats"
+
+	kitsync "github.com/rudderlabs/rudder-go-kit/sync"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	asynccommon "github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+	"github.com/rudderlabs/rudder-server/router/batchrouter/circuitbreaker"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+type PartitionWorker struct {
+	wg      sync.WaitGroup
+	logger  logger.Logger
+	brt     *Handle
+	channel chan *JobEntry
+	active  *config.Reloadable[int]
+	cb      circuitbreaker.CircuitBreaker
+	limiter kitsync.Limiter
+}
+
+type JobEntry struct {
+	job      *jobsdb.JobT
+	sourceID string
+	destID   string
+}
+
+func NewPartitionWorker(logger logger.Logger, partition string, brt *Handle, cb circuitbreaker.CircuitBreaker) *PartitionWorker {
+	maxJobsToBuffer := brt.conf.GetInt("BatchRouter.partitionWorker.maxJobsToBuffer", 100000)
+	pw := &PartitionWorker{
+		channel: make(chan *JobEntry, maxJobsToBuffer),
+		logger:  logger.With("partition", partition),
+		active:  brt.conf.GetReloadableIntVar(1, 1, "BatchRouter.partitionWorker.active"),
+		brt:     brt,
+		cb:      cb,
+		limiter: kitsync.NewLimiter(
+			context.Background(), &sync.WaitGroup{}, "batchrouter_partition_worker",
+			brt.conf.GetInt("BatchRouter.partitionWorker.concurrency", 10),
+			stats.Default,
+		),
+	}
+	return pw
+}
+
+func (pw *PartitionWorker) AddJob(job *jobsdb.JobT, sourceID, destID string) {
+	pw.channel <- &JobEntry{
+		job:      job,
+		sourceID: sourceID,
+		destID:   destID,
+	}
+}
+
+func (pw *PartitionWorker) Start() {
+	uploadFreq := pw.brt.uploadFreq.Load()
+	pw.logger.Infof("Starting partition worker with upload frequency %s", uploadFreq)
+	processJobs := func(jobs []*JobEntry) {
+		pw.wg.Add(1)
+		done := pw.limiter.Begin("")
+		go func() {
+			defer done()
+			defer pw.wg.Done()
+			jobsPerConnection := lo.GroupBy(jobs, func(job *JobEntry) lo.Tuple2[string, string] {
+				return lo.Tuple2[string, string]{A: job.sourceID, B: job.destID}
+			})
+			var wg sync.WaitGroup
+			for connection, jobs := range jobsPerConnection {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					pw.processAndUploadBatch(connection.A, connection.B, lo.Map(jobs, func(job *JobEntry, _ int) *jobsdb.JobT { return job.job }))
+				}()
+			}
+			wg.Wait()
+		}()
+	}
+	pw.wg.Add(1)
+	go func() {
+		defer pw.wg.Done()
+		for {
+			jobs, jobsLength, _, ok := lo.BufferWithTimeout(pw.channel, pw.brt.maxEventsInABatch, uploadFreq)
+			if jobsLength > 0 {
+				processJobs(jobs)
+			}
+			if !ok {
+				return
+			}
+		}
+	}()
+}
+
+func (pw *PartitionWorker) processAndUploadBatch(sourceID, destID string, jobs []*jobsdb.JobT) {
+	if len(jobs) == 0 {
+		return
+	}
+	// Get destination and source details
+	pw.brt.configSubscriberMu.RLock()
```

Comment:
no need to repeat same code twice, if we cannot find the source we should update the job status with the provided error
```go
	getSource := func(destID string, sourceID string) (backendconfig.SourceT, error) {
		pw.brt.configSubscriberMu.RLock()
		destWithSources, ok := pw.brt.destinationsMap[destID]
		pw.brt.configSubscriberMu.RUnlock()
		if !ok {
			return backendconfig.SourceT{}, fmt.Errorf("destination not found in config for id: %s", destID)
		}
		s, found := lo.Find(destWithSources.Sources, func(s backendconfig.SourceT) bool {
			return s.ID == sourceID
		})
		if !found {
			return backendconfig.SourceT{}, fmt.Errorf("source not found in config for id: %s", sourceID)
		}
		return s, nil
	}
```